### ISSUE TITLE
Enhancement: Marking item "read" via OwnCloud/Nextcloud slows down the interface significantly

### ISSUE THREAD

**Issue by groentebroer** [martinrotter/rssguard#71]
#### Brief description of the issue.

I am using an OwnCloud / NextCloud account with RSS Guard 3.3.5.

When scrolling through the items, it marks every item individually read. The next item only loads after the previous item has been read.
My NextCloud instance is rather slow, so it might take 2 seconds to respond. This means that the interface of RSS Guard also waits 2 seconds.
#### How to reproduce the bug?
1. Setup a Synced ownCloud/NextCloud account (together with the "news" plugin)
2. Skip through new "unread" items and you can see the "slowdown".
3. 
#### What is the expected result?

Browsing is done locally.
Two solutions I can think of right now;
1. Do the "mark read" on a different thread not affecting the interface.
2. Do the "mark read" only locally and when updating the feeds, update the "mark read" in batch.
#### What actually happened?

The interface seems to hang.

**Comment by martinrotter** [comment]
Yes, working with the server is done on the main GUI thread, thus blocking the GUI when the operation takes too long.

Next versions will only fix bugs and this is one of things I am thinking about fixing.

**Comment by martinrotter** [comment]
The reason of this being on GUI thread that the state of the messages is update in the local DB **after** it is updated in remote TT-RSS/owncloud server so that both message states (one in local db and one on the remote server) are the same.

But I guess I can make this asynchronous. The remote state will be updated first on the background and when the operation finishes, local changes will be made.

**Comment by groentebroer** [comment]
Yes, I do not get any notification anymore, because the updated number stays 0 ;-).
Refreshing a single feed (which I know that has updates via the log) gives me a notification of (in this case) 5 updated entries. (So it seems the DB lock is not "active" when updating a single feed.)

Hope this helps pinpointing the issue ;).

**Comment by martinrotter** [comment]
Practically, it is a bit weird to use option 2. Users who do not have auto-updates enabled would probably forget that they have to update feed to get cached message states pushed to sync server.

But, I feel that the best way to solve this might be this one perhaps:

> 1. Keep data structure which holds read/unread/starred states of messages you have changed.
> 2. From time to time, go through the structure and update states of those messages on the remote server with one (or two) HTTP queries.
> 3. When app quits, check data structure once more.

This will have two advantages:

1. When you go through messages or do any change to state of messages, then the change will be cached locally, which is very fast operation, thus no UI slowdown.
2. Changes are pushed to the server from time to time with less HTTP traffic.

**Comment by mahdi1234** [comment]
While not being original reporter I second this being really helpful ... in my opinion local instance should mark item as read immediately and different thread would send notification to tt server in background not locking GUI or as batch as you proposed in last comment ... actually batch may be even better as you know exact internal state of items and once in a while send updates to server, when internal state doesn't change no ping to server :)

**Comment by groentebroer** [comment]
I think handling it in batch with info you already know would be the best solution. (Like mahdi1234 also mentioned)
Then you also do not have to "still do something" when quitting.

**Comment by martinrotter** [comment]
SO, I just pushed EXPERIMENTAL fix. Fix so far fixes following:

 - only owncloud plugin,
 - only manual read/unread message changes.

Changes are cached and pushed to server each 30 seconds and when app exits. All this is done in separate thread, thus no GUI blocking is present.

**PLEASE, test this function ASAP:**
1. Get last RSS Guard development build (at least commit 09630f1).
2. Add owncloud account, update some feeds to get list of messages.
3. Go through unread messages with keyboard arrows, marking them one by one as read when hitting arrow buttons.
4. See the if the GUI response is okay.
5. Check if message read/unread state gets really updated on owncloud server by manually checking on owncloud server of possible.

So far, please, update feeds only once. When you update feed the cached metadata will **NOT** get saved yet, this will be added shortly. But first, please test that this simple scenario just works.

**Comment by martinrotter** [comment]
https://github.com/martinrotter/rssguard/commit/013a2ff0e625a4a4fbad054853b091c2b93de957

**Comment by groentebroer** [comment]
Hi Martin,

Wow, wow, wow! This speeds up things with an amazing magnitude!
I can confirm that reading the messages one at a time, and marking a whole feed read works in build 013a2ff. And as fast as "regular" feeds.

I also checked the network traffic and can see the update with multiple ID's periodically as you stated and I can see the changes within OwnCloud / Nextcloud News.

1 Thing that I noticed is that it always requests batchSize "-1" for all feeds, which results in all messages (I think). Maybe this can be optimized or limited to the latest 50? or something?

Thanks for this great fix, keep up the good work and hope you are doing better on your illness!

I am quite busy this weekend, but will use this build now for my OwnCloud sync if anything might arise, I will let you know!

**Comment by martinrotter** [comment]
https://github.com/martinrotter/rssguard/commit/49f76775f411b1b562d5f8ca7c9147b2f33f23aa

Please test.

**Comment by martinrotter** [comment]
1. Mark some messages read.
2. And then fast update that feed.
3. Check that messages you marked read are still marked read.

**Comment by martinrotter** [comment]
The `batch_size = -1` is used [here.](https://github.com/owncloud/news/wiki/Items-1.2#get-items) It tells the owncloud server to simply send all available messages for one particular feed at once. I figured out that this is faster then requesting those messages through several HTTP connections. At least it won't DDOS the server with multiple connections.

**Comment by groentebroer** [comment]
I will test this situation with build 49f7677 and will let you know my findings soon.

Maybe limit the batch-size at 50 or 100 or something, because now it is several Megabytes per feed for my feeds...

**Comment by martinrotter** [comment]
I see. Can you create new ticket for it?

**Comment by groentebroer** [comment]
Yes, Will do ;)

**Comment by groentebroer** [comment]
I can confirm that your test scenario is completed succsessfull.

Marking those items read and then updating the feed manually, results in the items being marked read to the server first, before updating the feed, which seems to work perfectly.

I also tried marking some items read, then the whole feed, or another hole feed, which also works well.

One thing I notice in the latest build is that marking a complete feed read sends all the ID's to the owncloud backend, instead of just the unread ones that should be changed to read. (I cannot remember exactly, but thought the previous build only updated the unread ones to read.)
It works well, but sending only the unread ones to mark read, will reduce the overhead.

**Comment by mahdi1234** [comment]
Is this also for tt-rss? I can try it then as well.

**Comment by martinrotter** [comment]
Not yet, the implementation uses base class, so it can be enabled for tt-rss in two minutes, I will do in a few days probably, dont worry. :)

**Comment by martinrotter** [comment]
OK, added experimental support for TT-RSS too. Wait for latest commit to compiler and test. Then let me know here pls.

https://github.com/martinrotter/rssguard/commit/858184b6ff7e4a2ed908c36480de9b25015e43d4

**Comment by mahdi1234** [comment]
So far only quick tests against tt-rss ... but I'm really happy it's blazing fast! checked with 858184b keep up the good stuff :)

**Comment by martinrotter** [comment]
OK. I will cache also marking of message starred/unstarred and I will review the solution, then I will mark this as closed if you feel it works.

**Comment by martinrotter** [comment]
So, please test TT-RSS to more detail if possible. I will then mark this as closed if all is okay.

**Comment by mahdi1234** [comment]
I am using 7f16ae5 with tt-rss for few days and no problems so far.

**Comment by martinrotter** [comment]
OK, closing this. Feel free to open new bug if there is any kind of regression.
